### ISSUE TITLE
Persist sys-vpn connection properties configured in NetworkManager

### ISSUE THREAD

**Issue by yourcase** [QubesOS/qubes-issues#8727]
### The problem you're addressing (if any)
sys-vpn comes with an automatically configured NetworkManager connection to a NetVM, which in my case shows up as "VM uplink eth0".
Properties like the following don't survive a restart of sys-vpn:
- Edit connections -> VM uplink eth0 -> edit -> General -> Automatically connect to VPN
- or: ... IPv4 Settings -> routes
"Saving" stores these settings in /rw/config/NM-system-connections/qubes-uplink-eth0, but that file gets overwritten on restart and the configuration is lost.

As a consequence I currently redo the configuration manually every time I restart sys-vpn.

### The solution you'd like
I'd like my configuration not to disappear.

### The value to a user, and who that user might be
The value would be not to have to redo manual configuration all the time, and thus have a connection that automatically connects to a VPN, and remembers a local route to say only a printer reachable from the lan.

**Comment by yourcase** [comment]
Not that I'm aware of. If I stop sys-vpn, it remains visible in the Qube Manager, which I take to mean it's not disposable.

**Comment by DemiMarie** [comment]
> > Is your sys-vpn a disposable?
> 
> Not that I'm aware of. If I stop sys-vpn, it remains visible in the Qube Manager, which I take to mean it's not disposable.

That could also be a named disposable.  Can you run `qvm-prefs sys-vpn klass`?

**Comment by yourcase** [comment]
`[screen@dom0] qvm-prefs sys-vpn klass`
`AppVM`

Just to be clear: If I enter IPv4-routes in the NetworkManager GUI, and press SAVE, then the routes show up in `/rw/config/NM-system-connections/qubes-uplink-eth0`, which is linked to from /etc. If I restart sys-vpn, they are not in that file any more, presumably because the network connection between sys-vpn and and its NetVM is set up from on high as if it was a manual configuration (i.e. no DHCP between the VMs). That "manual" configuration is then written into /rw/config and overwrites the route entries.

Is that reproducible on your side?

**Comment by marmarek** [comment]
`qubes-uplink-eth0` connection is generated on VM start, mostly to reflect settings like IP address, gateway, DNS etc. So, this issue shows we shouldn't override the whole file, but only update settings that needs to be updated.

**Comment by SolarSorcerer** [comment]
Is it a standalone qube? Some of the services might be affected if it is just a normal template based qube.

**Comment by yourcase** [comment]
> Is it a standalone qube?

My `sys-vpn` is based on a `fedora-38` template. It takes the standard `sys-firewall` as a Net qube. (I'm implicitly trusting the vpn provider). Sys-vpn's gateway IP address is `10.138.22.156`, which is `sys-firewall`'s IP address. Note that a typical lan route in `sys-vpn` will show up in `/rw/config/NM-system-connections/qubes-uplink-eth0` in the `[ipv4]` section like below:

```
[ipv4]
address1=10.137.0.48/32,10.138.22.156
dns=10.139.1.1;10.139.1.2;
may-fail=false
method=manual
route1=192.168.22.1/32,10.138.22.156
```

Meaning that access to, say, a printer at `192.168.99.137` will bypass the vpn tunnel, and instead to directly to `10.138.22.156`, which is `sys-firewall` which gives access to the lan.

It would of course be super handy if your enhancement would pick up the fact that `10.138.22.156` is in fact a VM internal to qubes, so that if qubes decides to change `sys-firewall`'s IP address, it would then adjust this route accordingly. I haven't checked if that IP address changes frequently. If it doesn't then it would just be a nice to have. If on the other hand it changes every time `sys-firewall` restarts, then such an automatic route update would save me frequent manual reconfiguration. I don't think I want to route directly to the lan gateway outside of qubes because I suspect that increases the leak potential to the internet, since any process running in a VM behind `sys-vpn` would then be able to directly access a gateway to the internet bypassing the vpn, correct me if I'm wrong.

Anyway, just having the routes survive the `sys-vpn` qube restart would already be a great improvement!

**Comment by SolarSorcerer** [comment]
So, to triple check it is a standalone VM? It must be a standalone VM in your use case.

**Comment by yourcase** [comment]
> standalone VM?

I don't know how to verify that, I'm appending the output of `qvm-prefs sys-vpn` in `dom0` :
[qvm-prefs_sys-vpn_out.txt](https://github.com/QubesOS/qubes-issues/files/13497647/qvm-prefs_sys-vpn_out.txt)

**Comment by DemiMarie** [comment]
NetworkManager connections should be stored in a persistent directory.  If not, that’s a bug somewhere.

**Comment by marmarek** [comment]
> `qubes-uplink-eth0` connection is generated on VM start, mostly to reflect settings like IP address, gateway, DNS etc. So, this issue shows we shouldn't override the whole file, but only update settings that needs to be updated.

I think this my comment was missed in the above discussion.
Let me point at the specific code place: https://github.com/QubesOS/qubes-core-agent-linux/blob/main/network/setup-ip#L105

**Comment by DemiMarie** [comment]
> > `qubes-uplink-eth0` connection is generated on VM start, mostly to reflect settings like IP address, gateway, DNS etc. So, this issue shows we shouldn't override the whole file, but only update settings that needs to be updated.
> 
> I think this my comment was missed in the above discussion. Let me point at the specific code place: https://github.com/QubesOS/qubes-core-agent-linux/blob/main/network/setup-ip#L105

Should we use `nm-cli` instead of creating the file ourselves?

**Comment by SolarSorcerer** [comment]
When you created the sys-vpn, did you make it a standalone qube?
And it should say it in qube manager under the template column. See attatched image.
![12308f2](https://github.com/QubesOS/qubes-issues/assets/109770246/b6b6f6f4-4eaf-4007-a3a9-a3880b80e3f3)

Also, if you don't know how app and standalone qubes work here is a summary.
```
App Qube: Root is amnestic, Home isn't. Meaning only the root will reset back to what it template's root is whenever it restarts.
Standalone VM: Root and home aren't amnestic. Is completely standalone so it doesn't reset anything.
```

**Comment by marmarek** [comment]
> Should we use `nm-cli` instead of creating the file ourselves?

I believe the script runs before starting NetworkManager, so that probably won't work.

**Comment by DemiMarie** [comment]
> > Should we use `nm-cli` instead of creating the file ourselves?
> 
> I believe the script runs before starting NetworkManager, so that probably won't work.

The only other option I can think of is to edit the files in the script, which I am _really_ not interested in doing.  Writing a file is one thing, _changing_ an existing file (which could have who knows what in it) is another.  Is there a reason we can’t just start NetworkManager in the script?

**Comment by marmarek** [comment]
> Is there a reason we can’t just start NetworkManager in the script?

It will immediately automatically create connection (with DHCP enabled) and will try to use it (including when you later dynamically want to change netvm)

**Comment by DemiMarie** [comment]
What about using `NM_UNMANAGED` in a udev rule to mark the interface as not managed by NetworkManager?

**Comment by DemiMarie** [comment]
The safest way to prevent leaks is to use explicit nftables rules set up in a systemd service required by `network-pre.target`.  However, this is a manual process involving user-unfriendly command-line tools, so it isn’t a good solution here.

`man nm-settings-keyfile` indicates that `GKeyFile` (from GLib) is used, so this can be solved by using GLib to edit the configuration.  I’m not confortable trying to do the same in a shell script.

**Comment by yourcase** [comment]
> as simple as a modification of script X

Ah, I see from @marmarek 's code-pointer that script X is a bash script. Maybe port to python and use an ini-file read/write library? Something like this: https://docs.python.org/3/library/configparser.html ?

**Comment by SolarSorcerer** [comment]
![IMG_20231130_095730.jpg](https://github.com/QubesOS/qubes-issues/assets/109770246/beb0beac-9239-45a5-8c44-5f5495d19190)

Its a AppVM so any modifications you are making to it outside if the home directory will he reset.

So I suggest either making it standalone or create a new template VM with your modifications.

**Comment by marmarek** [comment]
First, just using standalone qube won't help, because the script above. But then, the "system-connections" dir is [stored on /rw](https://github.com/QubesOS/qubes-core-agent-linux/blob/main/network/network-manager-prepare-conf-dir) so it will persist.

> @marmarek : Are there any unit tests for https://github.com/QubesOS/qubes-core-agent-linux/blob/main/network/setup-ip ?

No, but there are [integration tests](https://www.qubes-os.org/doc/automated-tests/#integration--unit-testing-in-dom0). We have it integrated with openQA, so you can open a PR you believe it should work and just ask to include in the next test run.

**Comment by MalKeshar** [comment]
FYI, I know of one legitimate case where /etc/NetworkManager/system-connections/qubes-uplink-eth0 disappears.
When using Netplan with NetworkManager, the system moves the connection configuration to /etc/netplan/.
I’ve seen this in the Ubuntu 24.04 template.

I ended up setting up the connection manually.
I believe it should be possible to simply disable Netplan, but I haven’t tested that option.
